### DISCUSSION TITLE
Revisiting random seeds

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#96] [Q&A] [answered]
@shikokuchuo, thanks to your work on `secretbase`, `targets` now has a robust solution for reproducible parallel-safe random numbers. I am starting this discussion to see how this could improve `mirai` and/or `crew`. Both currently use L'Ecuyer as a default, and according to `?RNGkind`, it "exhibits 6 clear failures in each of the TestU01 Crush and the BigCrush suite (L'Ecuyer, 2007)". What are your preferences for `mirai`?

I am thinking about switching `crew` away from the `rs` argument of `mirai::daemon()` and instead having `crew::crew_worker()` call `set.seed(kind = "default", seed = secretbase::sha3(x = list(launcher, worker, instance, nanonext::mclock()), bits = 32L, convert = NA))` before it calls `mirai::daemon()`. Would I need to do anything else to make sure `mirai` does not change the algorithm back to L'Ecuyer?

**Comment by shikokuchuo** [comment]
I'm mindful of https://github.com/ropensci/targets/issues/1139#issuecomment-1906908495 I think Henrik's recommendation is spot on conceptually - hence as `mirai` already implements L'Ecuyer streams, it seems useful to allow them. 

As for setting the seed in `crew`, if you intend to do that in `crew::crew_worker()` before you call `mirai::daemon()` that would be fine.  Just make sure you are not specifying 'rs' in that case as otherwise `mirai::daemon()` will assign that to `.Random.seed` at the start of the call.

As for what a default should be, I'll need further time to consider. Using `secretbase` would allow seeds to be set independently and move to MT as you point out.

**Reply by shikokuchuo** [reply, marked as answer]
Practically, I don't think it's any more cumbersome to call `mirai::nextstream()` rather than `secretbase::sha3()`.

Cet par, you'd prefer MT over L'ecuyer due to those failures. However, in the parallel computing context, the overlapping RNG ranges are, if not a bigger concern, at least one that is more narrowly defined for the domain our software has been written for. Hence I think ~~logically~~ _from a human psychology perspective_ it is this risk that we have to address foremost. Even without looking empirically at which is more severe.

Of course, the targets-type reproducibility requirement means that it's a no-brainer to use `secretbase::sha3()` there, but at least for `mirai` there isn't the assumption that everyone is going to need reproducibility.

Does that make sense?

**Reply by wlandau** [reply]
Yes, I think this is well-reasoned. That makes sense.
